USEID-158 Add ADR for API keys (#41)
digitalservicebund/useid-backend-service: doc/adr/0009-api-key.md

# 9. API Keys

Date: 2022-07-06

## Status

Accepted

## Context

> We want to filter incoming requests in the UseID backend that only client with an approved integration could access UseID endpoints.

Reviewer: ```suggestion
We want to filter incoming requests in the UseID backend, so that only authorized clients are able to access the API.
```

## Decision

> We use Bearer authentication. The Client (eService) should add a token in the Authorization header when requesting to the endpoints and the token should be a random string of 51 characters.

Reviewer: ```suggestion
We use Bearer authentication. The Client (eService) should add a token in the Authorization header when requesting to the endpoints and the token should be a string of 51 characters which was issued to the client by us.
```

e.g.

`Authorization: Bearer <token>`

> The token should, for MVP purpose, be configured as environment variable. For the future releases, a lifecycle of token should be defined, e.g. validity of the token.

Reviewer: I would remove the whole sentence since it is an implementation detail which is not relevant for the decision.

## Consequences

> 1. Reduce our security risk, since we only allow known client in the UseID backend

Reviewer: ```suggestion
1. Reduce our security risk, since we only allow authorized clients to access the API
```

> 2. Reduce resource usage, since we are filtering unknown requests, and we do not process them further

Reviewer: Not relevant. Would remove. I think the API key is for security reasons only.
